include/sparrow/memory.hpp

```diff
@@ -0,0 +1,94 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or mplied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cassert>
+#include <memory>
+
+namespace sparrow {
+
+/**
+ * @brief A value_ptr is a smart pointer that behaves like a value.
+ * It manages the lifetime of an object of type T.
+ * When copied, it copies the managed object.
+ * 
+ * @tparam T The type of the object managed by the `value_ptr`.
+ */
+template <class T>
+class value_ptr {
+public:
+    value_ptr() = default;
+
+    explicit value_ptr(T value)
+        : value_(std::make_unique<T>(std::move(value))) {}
+
+    explicit value_ptr(T* value) 
+        : value_(value != nullptr ? std::make_unique<T>(*value) : std::unique_ptr<T>()) {}
+
+    value_ptr(const value_ptr& other)
+        : value_(other.value_ ? std::make_unique<T>(*other.value_) : std::unique_ptr<T>()){}
+
+    value_ptr(value_ptr&& other) noexcept = default;
+
+    ~value_ptr() = default;
+
+    value_ptr& operator=(const value_ptr& other) {
+        if (other.has_value()) {
+            *value_ = *other.value_;
```

Comment:
```suggestion
            if(value_)
            {
                *value_ = *other.value_;
            }
            else
            {
                value_ = std::make_unique(*other.value_);
            }
```

The code before my suggestion is ub if `value_` is `nullptr`, you need to check for it too and only `if(value_ && other.has_value())` can you do the copy of pointed values. If `not value_` then you need to do the code you did previously with make_unique of the value of other.

Reply:
Good catch

Reply:
I added UT too